### PR TITLE
namespace controller refactor

### PR DESCRIPTION
**Pull request by raz-bn**

### CODE REVIEW

File: controllers/namespace_controller.go [hunk 1]
```diff
@@ -108,58 +117,50 @@ func (r *NamespaceReconciler) Update(ctx context.Context, log logr.Logger, names
...
 	}
-	//sns is gone did we already remove the finalizer
-	if !controllerutil.ContainsFinalizer(namespace, danav1alpha1.NsFinalizer) {
-		return false, nil
+
+	if err := r.Delete(ctx, &sns); err != nil {
+		//We should treat the case when the Delete did not find the subnamespace we are trying to delete
+		//since we must make sure the finalizer is removed even though the subnamespace is missing
+		if apierrors.IsNotFound(err) {
+			goto removeFinalizer
```
Comment (RazcoDev) [inline controllers/namespace_controller.go]: goto can be useful but here it makes your error handling messier.

File: controllers/namespace_controller.go [hunk 2]
```diff
@@ -108,58 +117,50 @@ func (r *NamespaceReconciler) Update(ctx context.Context, log logr.Logger, names
...
+//CleanUp is being called when a namespace is being deleted,it deletes the subnamespace object related to the namespace inside its parent namespace,
+//also removing the finalizer from the namespace so it could be deleted
+func (r *NamespaceReconciler) CleanUp(ctx context.Context, log logr.Logger, namespace *corev1.Namespace) error {
+
+	sns := danav1alpha1.Subnamespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      namespace.Annotations[danav1alpha1.SnsPointer],
+			Namespace: namespace.Labels[danav1alpha1.Parent],
+		},
 	}
```
Comment (RazcoDev) [inline controllers/namespace_controller.go]: Can be on a seperated object-builder function

File: controllers/namespace_controller.go [hunk 3]
```diff
@@ -108,58 +117,50 @@ func (r *NamespaceReconciler) Update(ctx context.Context, log logr.Logger, names
...
 		role = danav1alpha1.Leaf
 	}
-	//if ns has no parent then he is a root
-	if _, ok := namespace.Labels[danav1alpha1.Parent]; !ok {
-		role = danav1alpha1.Root
-	}
 
-	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, namespace, func() error {
-		namespace.Annotations[danav1alpha1.Role] = role
+	if namespace.Labels[danav1alpha1.Parent] == role {
```
Comment (RazcoDev) [inline controllers/namespace_controller.go]: labels parent == role ?
two different types of labels ?

File: controllers/namespace_controller.go [hunk 4]
```diff
@@ -48,57 +49,65 @@ type NamespacePredicate struct {
...
+
+	controllerutil.AddFinalizer(namespace, danav1alpha1.NsFinalizer)
+	if err := r.Update(ctx, namespace); err != nil {
 		if apierrors.IsConflict(err) {
-			log.Info("newer resource version exists")
+			log.V(2).Info("newer resource version exists")
 			return nil
 		}
-		log.Error(err, "unable to add finalizer")
+		log.Error(err, "unable to update namespace")
```
Comment (RazcoDev) [inline controllers/namespace_controller.go]: log.V() convention missing

File: controllers/namespace_controller.go [hunk 5]
```diff
@@ -48,57 +49,65 @@ type NamespacePredicate struct {
...
 	var ns corev1.Namespace
 
+	//get namespace instance
 	if err := r.Get(ctx, req.NamespacedName, &ns); err != nil {
 		if apierrors.IsNotFound(err) {
-			log.Info("deleted")
+			log.V(2).Info("deleted")
 			return ctrl.Result{}, client.IgnoreNotFound(err)
 		}
 		log.Error(err, "unable to get Namespace")
```
Comment (RazcoDev) [inline controllers/namespace_controller.go]: log.V convention missing ?

File: controllers/namespace_controller.go [hunk 6]
```diff
@@ -108,58 +117,50 @@ func (r *NamespaceReconciler) Update(ctx context.Context, log logr.Logger, names
...
+	if namespace.Labels[danav1alpha1.Parent] == role {
 		return nil
-	}); err != nil {
-		log.Error(err, "unable to update role")
-		return err
 	}
 
+	namespace.Annotations[danav1alpha1.Role] = role
+	if err := r.Update(ctx, namespace); err != nil {
+		log.Error(err, "unable to update namespace")
```
Comment (RazcoDev) [inline controllers/namespace_controller.go]: log.V convention missing

File: controllers/namespace_controller.go (hunk 1 above)
Comment (raz-bn) [reply controllers/namespace_controller.go]: do you have an idea how to deal with the case where 
1. sns deleted 
2. remove finalized failed -> return err and reconcile again 
4. sns not found - > return err and reconcile again (loop)

also notFound should not be an error, maybe info

File: controllers/namespace_controller.go (hunk 3 above)
Comment (raz-bn) [reply controllers/namespace_controller.go]: Good catch, should be danav1alpha1.Role

File: controllers/namespace_controller.go (hunk 4 above)
Comment (raz-bn) [reply controllers/namespace_controller.go]: What should log level errors be at?

**Pull request merged**